Add Pyodide/WASM support in CI for `numcodecs`

agriyakhetarpal opened the pull request:
This PR adds a CI job (`emscripten-ci.yaml`) that builds WASM wheels for numcodecs for use in Pyodide-based environments using the Emscripten toolchain. This was discussed in one of the [community meetings for Zarr](https://hackmd.io/tFcffc__QRuo1UdGcGUbsg#2024-03-06) where it was said that this change would be welcome as a pull request, and follow-up work after this is complete is planned:

1. A job that publishes nighty WASM wheels for numcodecs to a PyPI-like distribution channel (not sure where, this might require requesting access on the https://github.com/scientific-python/upload-nightly-action repository (which corresponds to https://anaconda.org/scientific-python-nightly-wheels/)
2. A corresponding job for out-of-tree testing for Zarr that builds up on this job (and subsequently its nightly pushes)
3. Interactive documentation improvements with Sphinx that make use of these nightly wheels in some form (longer-term goal, soon but not very soon I guess)

TODO:

- [x] Unit tests and/or doctests in docstrings
- [x] Tests pass locally
- [ ] Docstrings and API docs for any new/modified user-facing classes and functions
- [x] Changes documented in docs/release.rst
- [ ] Docs build locally
- [x] GitHub Actions CI passes
- [x] Test coverage to 100% (Codecov passes)

### Additional context

Some xrefs for reviewers and project maintainers:

- NumPy's out-of-tree job: https://github.com/numpy/numpy/blob/main/.github/workflows/emscripten.yml and corresponding PR: https://github.com/numpy/numpy/pull/25894
- The same for PyWavelets (optional dependency for `scikit-image`): https://github.com/PyWavelets/pywt/pull/701 and https://github.com/PyWavelets/pywt/pull/714
- For the `pandas` repository https://github.com/pandas-dev/pandas/pull/57896
- https://github.com/scikit-image/scikit-image/pull/7350
- and so on

agriyakhetarpal committed:
Git-ignore pytest cache folders

agriyakhetarpal committed:
Refine outdated sections for Contributing guide

This commit makes the following changes:

1. Formatting and whitespace changes
2. Removes outdated Python version references
and adds references to the current supported
version of Python for NumCodecs
3. Removes references to Travis and AppVeyor
since GitHub Actions is being used in favour of
them as a CI provider.

agriyakhetarpal committed:
Add patch to disable threading and multiprocessing

agriyakhetarpal committed:
Add patch to embed missing POSIX syscall headers

agriyakhetarpal committed:
Add initial workflow to run + debug Pyodide builds

agriyakhetarpal committed:
Disable AVX2 and SSE2 instructions

agriyakhetarpal committed:
Use separate build and test jobs and share wheel

agriyakhetarpal committed:
Add patch that allows loading `.npy` files

agriyakhetarpal committed:
Apply patch for Emscripten to load `.npy` files

agriyakhetarpal committed:
Add a constant to check if platform is WASM

agriyakhetarpal committed:
Use `is_wasm` to disable multiprocessing tests

agriyakhetarpal committed:
Apply `.npy` files patch again when running tests

agriyakhetarpal committed:
Remove patch that is unneeded at build time

agriyakhetarpal commented:
The workflow is currently going to fail because the file system access in Emscripten has a bug with saving `.npy` files – I had started on this PR roughly two weeks back, and I have now rebased on top of the changes from the `main` branch. Here is the related issue that I opened on the Pyodide repository a few hours ago: https://github.com/pyodide/pyodide/issues/4779

@rgommers and I had fixed a similar error in https://github.com/PyWavelets/pywt/pull/701 by converting all of the `.npy` files present to `.npz` files. If this is a suitable workaround at this time and does not render other PRs un-mergeable, please let me know and I would be happy to apply this fix here (it works/worked for loading `.npz` files, it should most likely work for saving them as well).

joshmoore commented:
Thanks, @agriyakhetarpal. I've launched the workflows.

agriyakhetarpal commented:
Ah, I think the `miniconda` action needs updating. I'll push a commit to fix that in a moment.

agriyakhetarpal commented:
Never mind, looks like #528 is already on it – I can wait and rebase later after that gets merged. For the relevant jobs here, [the Pyodide wheel build succeeds](https://github.com/zarr-developers/numcodecs/actions/runs/9180978236/job/25266750320?pr=529) while [the tests fail](https://github.com/zarr-developers/numcodecs/actions/runs/9180978236/job/25266967311?pr=529) because of the reasons described above, and the rest of the test suite passes.

agriyakhetarpal commented:
Marking as a draft till there is a fix for this on the Pyodide and Emscripten ends. I will put my focus on an accompanying CI job and associated fixes for the Zarr repository.

agriyakhetarpal committed:
Merge branch 'main' into setup-emscripten-ci

agriyakhetarpal committed:
Delete patch that's not needed now

agriyakhetarpal committed:
Ignore Pyodide cross-build environment files

agriyakhetarpal committed:
Revamp workflow and clean up changes

agriyakhetarpal commented:
It's been a while since this PR has been opened, could we get the ball rolling by temporarily skipping the failing test for now? I feel that `numcodecs`'s CI can benefit from adding this workflow for the rest of the test suite, at least. I've revisited the merge conflicts in the previous few commits and updated the workflow file in 764f34e5c0fd1c31a5316af760e3fc1848e28410, I'll see how the builds go.

agriyakhetarpal commented:
It's been a while since this PR has been opened, could we get the ball rolling by temporarily skipping the previously failing test for now? I feel that `numcodecs`'s CI can benefit from adding this workflow for the rest of the test suite, at least. I've revisited the merge conflicts in the previous few commits and updated the workflow file in 764f34e5c0fd1c31a5316af760e3fc1848e28410, I'll see how the builds go.

agriyakhetarpal committed:
Fix linter errors and style failures

agriyakhetarpal committed:
Catch and disable `mutex` assignment

agriyakhetarpal committed:
Fix patch that disables multiprocessing, threading

agriyakhetarpal committed:
Don't hardcode `zlib` paths in patch details

agriyakhetarpal committed:
Install test + runtime extras, disable pytest cache

agriyakhetarpal committed:
Fix shell expansion in installation command

agriyakhetarpal committed:
Temporary: skip tests that require crc32c

agriyakhetarpal committed:
Merge branch 'main' into setup-emscripten-ci

agriyakhetarpal commented:
I've created https://github.com/pyodide/pyodide/pull/5169 and https://github.com/pyodide/pyodide/pull/5172 (please see above) for adding some missing dependencies to Pyodide. As a temporary measure, I've skipped the sole test that requires an updated version of the `crc32c` package: `numcodecs/tests/test_checksum32.py::test_crc32c_checksum` in 53f2096a5e8ae3b3f9fcc8779b1d6525291e66ca. I'll mark this PR as ready for review for now. @joshmoore, could you please trigger the workflows for me again?

agriyakhetarpal committed:
Temporary: skip `crc32c` dependency until available

agriyakhetarpal committed:
Update ci-emscripten.yaml

agriyakhetarpal committed:
Fix zarr installation order to collect more tests

agriyakhetarpal committed:
Force install zarr version 3 for CI

agriyakhetarpal committed:
Skip a test that requires threads

agriyakhetarpal committed:
Temporary: skip Zarr's async functionality tests

agriyakhetarpal committed:
Guard import of `multiprocessing`

agriyakhetarpal committed:
Revert "Guard import of `multiprocessing`"

This reverts commit bf4249191e538c9d8462eef168ebed643e2c241d.

agriyakhetarpal committed:
Bump verbosity when running test suite

agriyakhetarpal committed:
Skip codec entrypoint test that requires processes

agriyakhetarpal committed:
Let `pytest` inherit config for doctests

agriyakhetarpal committed:
Revert "Let `pytest` inherit config for doctests"

This reverts commit a5907859ac2350cb65ea82ff9547798cde4ff8a7.

agriyakhetarpal committed:
Ignore a doctest that uses threads within the file

agriyakhetarpal commented:
https://github.com/agriyakhetarpal/numcodecs/actions/runs/11746955955 via my fork is the first passing workflow run, with a few caveats:

1. A (one) `crc32c` test is temporarily disabled (see above, will be available in the next Pyodide release)
2. `zfpy` is not available yet (see above, will be available in the next Pyodide release)
3. Zarr v3 tests are temporarily disabled until we upgrade to Zarr v3 in Pyodide, hopefully by the next release
4. I had to skip a doctest that was using threads in a messy manner, but it should work.

The rest of the skips/xfails are the usual WASM limitations where the `multiprocessing`/`threading` modules are not available, and we have a pretty significant portion of the test suite passing without issues. I've compiled `crc32c` and `zfpy` with Emscripten locally in the time being and triggered the test suite for `numcodecs` with them, which didn't reveal anything problematic. I'm open to including them in the CI job temporarily if needed – they don't take too long to compile.

agriyakhetarpal committed:
Merge `main`

agriyakhetarpal committed:
Fix skip reason error message for entrypoint test

agriyakhetarpal committed:
`crc32c` >=2.7 is now available with Pyodide 0.27.0

agriyakhetarpal committed:
Bump to Pyodide version 0.27.0

agriyakhetarpal committed:
Clean up changes

agriyakhetarpal commented on .github/workflows/ci-emscripten.yaml:
```diff
@@ -0,0 +1,97 @@
+          # For tests in test_zarr3.py
+          pip install zarr==3.0.0b1
+
+          # Install the built numcodecs WASM wheel and relevant dependencies
+          pip install $(ls dist/*.whl)"[msgpack,crc32c,test,test_extras]"
+          # TODO: get zfpy built in Pyodide and install it here
```
`zfpy` has been built in Pyodide, but against NumPy v2 (as described above). There is a chance that the tests should work with NumPy >=2, too if I were to install it separately instead of using the `[zfpy]` extra. I'm open to test that out if asked.

agriyakhetarpal commented on .github/workflows/ci-emscripten.yaml:
```diff
@@ -0,0 +1,97 @@
+
+name: Pyodide CI
+
+on:
+  # TODO: refine after this is ready to merge
+  [push, pull_request, workflow_dispatch]
+
```
```suggestion
on: [push, pull_request]

concurrency:
  group: ${{ github.workflow }}-${{ github.ref }}
  cancel-in-progress: true

```

agriyakhetarpal reviewed (commented):
Update from my previous comment above: I've cleaned up and merged the latest changes from the `main` branch, and [agriyakhetarpal/numcodecs/actions/runs/12572995519/job/35045552245](https://github.com/agriyakhetarpal/numcodecs/actions/runs/12572995519/job/35045552245) via my fork passes all tests. We released https://github.com/pyodide/pyodide/releases/tag/0.27.0 yesterday, and I've bumped to it here.

- `crc32c` was upgraded to version 2.7.1 in https://github.com/pyodide/pyodide/pull/5169 to fulfill the requirement, and is now available as a part of Pyodide 0.27.
- I did add `zfpy` in https://github.com/pyodide/pyodide/pull/5172, but it appears that an upper bound on NumPy has been added here because it doesn't support it yet (https://github.com/LLNL/zfp/issues/210). Though, FWIW, via https://github.com/pyodide/pyodide/pull/4925, we bumped to NumPy v2 and did not find any problems with building and/or testing `zfpy` against it in specific. Unfortunately, we don't have support for multiple versions per package yet in the Pyodide distribution since it contains the distribution and all the WASM wheels together, so the `zfpy`-specific test(s) will have to remain disabled. Work is in progress to unvendor the packages from the Pyodide runtime and is one of our goals with Pyodide 0.28 later in the year.

Overall, this is the result from the test suite: `786 passed, 49 skipped, 36 xfailed in 23.22s`, which does seem quite promising.

Gentle ping, @joshmoore!

agriyakhetarpal committed:
Merge branch 'main' into setup-emscripten-ci

agriyakhetarpal commented:
Returning to this, I opened a PR to add `pcodec` to the Pyodide distribution, linked above. I hope we get it in a Pyodide 0.27.3 release sooner or later. I'm not aware of any new upcoming dependencies for `numcodecs` in the ambit, but we can get all of them added as long as they don't take too long to build.

joshmoore commented on docs/contributing.rst:
```diff
@@ -111,7 +111,7 @@ Creating a branch
 report the bug or propose the feature you'd like to add.
 
-It's best to create a new, separate branch for each piece of work you want to do. E.g.::
+It's best to create a new, separate branch for each piece of work you want to do. E.g.:
```
```suggestion
It's best to create a new, separate branch for each piece of work you want to do. E.g.::
```
The extra colon is intended for code-style quoting.

joshmoore commented on numcodecs/zarr3.py:
```diff
@@ -25,6 +25,19 @@
+# compiled via Emscripten where threads are not available, and accessing
+# the pytest config has no effect and ignoring warnings does not work.
+try:
+    import pytest
```
What would it take to not import pytest here?

agriyakhetarpal replied:
Last time I checked, we would need a way to run this doctest and not raise a warning that gets converted to an error. The other way could be to not add this doctest here, which is not the best way to proceed.

joshmoore reviewed (commented):
> [Agriya Khetarpal](https://github.com/agriyakhetarpal) commented [on Jan 1](https://github.com/zarr-developers/numcodecs/pull/529#pullrequestreview-2526907098)
> Update from my previous comment above: I've cleaned up and merged the latest changes from the main branch, and [agriyakhetarpal/numcodecs/actions/runs/12572995519/job/35045552245](https://github.com/agriyakhetarpal/numcodecs/actions/runs/12572995519/job/35045552245) via my fork passes all tests. We released [pyodide/pyodide@0.27.0 (release)](https://github.com/pyodide/pyodide/releases/tag/0.27.0) yesterday, and I've bumped to it here.

I have to admit this went under my radar on New Year's, @agriyakhetarpal. My apologies. But this is sounding amazing. Thanks so much for all the hard work here.

I've relaunched the workflows but since this is a new GHA ultimately I think we are going to have to get it in to evaluate.

I don't have any strong feelings on your recent comments (cc: @jakirkham @dstansby). Instead, I do feel that we should get this in and running sooner rather than later and we can always (partially) rollback if needed.

agriyakhetarpal committed:
Bring back code-style quoting

Co-authored-by: Josh Moore <josh@openmicroscopy.org>

agriyakhetarpal commented:
No worries at all, @joshmoore, and I'm grateful for the review! This would need to wait slightly until we get `pcodec` available (maybe later this month), and I expect that things will pass here without many problems. The doctest running in WASM is a tricky problem indeed – one way could be to not run the tests from the `docs/` directory, but as `numcodecs` doesn't follow the `src/` format, we have no option but to modify the `PYTHONPATH` or similar so that it doesn't try to import from the wrong `numcodecs/` directory.

agriyakhetarpal commented:
`pcodec` has been added to Pyodide via pyodide/pyodide#5432. I'll convert this PR to a draft one until we have it in a Pyodide release.

agriyakhetarpal committed:
Update to Pyodide 0.27.3, add `pcodec`

agriyakhetarpal commented:
We've cut a [`0.27.3` release](https://pyodide.org/en/stable/project/changelog.html#version-0-27-3) today, which bundles `pcodec`; all tests pass on my fork: https://github.com/agriyakhetarpal/numcodecs/actions/runs/13545498292?pr=529

agriyakhetarpal commented:
It looks like #569 is going to change a few things here – I have to note that Pyodide will need to set the provisioned `NUMCODECS_USE_SYSTEM_LIBS` variable to `OFF`, as we cross-compile from Linux to `wasm32-emscripten` and two patches are being applied to Blosc, for which we have to build it from the vendored sources.

I'm still happy to discuss with upstream maintainers if they're interested in such patches as I mentioned above in this PR, and I assume it helps both them and `numcodecs` that the patches aren't big at all.

joshmoore commented:
(Relaunched)

agriyakhetarpal committed:
Fix drop in coverage

agriyakhetarpal commented:
Thanks, I fixed the drop in the coverage – it's just that Codecov doesn't know about this WASM environment setup, so it can't be aware that this code path is being tested elsewhere.

agriyakhetarpal committed:
Fix another drop in coverage

agriyakhetarpal committed:
Merge branch 'main' into setup-emscripten-ci

agriyakhetarpal committed:
Don't add `pthread` for Emscripten

agriyakhetarpal committed:
Get version properly using `fetch-tags: true`

agriyakhetarpal committed:
Add CHANGELOG entry for Pyodide builds

agriyakhetarpal committed:
Wait for Git 2.48.1 for using `fetch-tags: true`

joshmoore commented:
Re-running jobs after:

```
Exception: Request failed after too many retries. URL: https://ingest.codecov.io/upload/github/zarr-developers::::numcodecs/upload-coverage
[PYI-4825:ERROR] Failed to execute script 'main' due to unhandled exception!
```

agriyakhetarpal committed:
Merge main

agriyakhetarpal commented:
I've addressed the merge conflicts introduced in the release notes file at the time of the 0.16 release. Thanks!

I wanted to note that there is a recent PR at #737, which also concerns the `wasm32-unknown-emscripten` target. The reason why the Pyodide CI job here can bypass that and work despite those changes is because [we drop a curated list of unsupported arguments when cross-compiling](https://github.com/pyodide/pyodide-build/blob/71a6d09651814439feb161537888078f7ff68c19/pyodide_build/pywasmcross.py#L234-L253). However, #737 will be a helpful change indeed.

agriyakhetarpal committed:
Merge main

agriyakhetarpal committed:
Ensure reliability of builds and tests

agriyakhetarpal committed:
`zfpy` is now available in Pyodide

agriyakhetarpal committed:
`zarr3` should also work now

agriyakhetarpal committed:
Ignoring Pyodide xbuildenv is no longer needed

agriyakhetarpal committed:
Bump to Pyodide 0.27.6, and Node 22

agriyakhetarpal commented:
Hi! I've resolved the new merge conflicts that had sprung up here. Also, I made sure that we're testing against the same Pyodide version as we're building the WASM wheels for. If you'd like me to, I can also try against the latest version of Pyodide to check for any regressions (which will use Emscripten 4.0.9 instead of 3.1.58) – cibuildwheel v3 soon will help doing this (i.e., testing against nightly or alpha versions of Pyodide), and a beta (3.0.0b1) is already out for testing.

Additionally, I enabled Zarr v3 as a dependency (it uses threads which won't work in WASM, so perhaps it makes sense to not add it or skip any new tests that will fail due to it). `zfpy` is also available in Pyodide 0.27.6 (which I've updated to, including a bump to Node 22), so that helps us not skip those tests. I had added a WASM build for `pcodec` and enabled it some time earlier.

agriyakhetarpal commented:
Test suite passes at https://github.com/zarr-developers/numcodecs/actions/runs/15199151699/job/42749864404?pr=529.

agriyakhetarpal commented:
This PR's sister PR, which adds Pyodide support for Zarr, is now also ready for review (marked above). Thank you!

agriyakhetarpal commented:
The sister PR to accompany this PR, which adds Pyodide support for Zarr, is now also ready for review (marked above). Thank you!

agriyakhetarpal committed:
Merge main*

agriyakhetarpal committed:
Modernise Emscripten workflow, build for 2026 ABI

agriyakhetarpal committed:
Fall back to `numpy.get_include()` for WASM

agriyakhetarpal committed:
Add Pyodide-related config for cibuildwheel

agriyakhetarpal committed:
Fix `is_wasm` import

agriyakhetarpal committed:
Skip pcodec tests

agriyakhetarpal committed:
Use `pytest.importorskip`, more modern test skips

agriyakhetarpal committed:
xfail two tests with size overflows for now

agriyakhetarpal committed:
Try to emit zarr3 deprecation warnings

agriyakhetarpal committed:
Fix zfpy and lzma skips

agriyakhetarpal committed:
Wrong tag is getting picked up, fetch-depth needed

agriyakhetarpal committed:
Remove workaround for deprecation warnings

agriyakhetarpal committed:
Streaming decompression tests pass now

agriyakhetarpal committed:
Discard changes to tests/test_zfpy.py

agriyakhetarpal commented:
The WASM build and test suite passes at https://github.com/zarr-developers/numcodecs/actions/runs/25944005084/job/76268019204?pr=529. Looks like there is an ongoing effort to upgrade pcodec to v1 in https://github.com/zarr-developers/numcodecs/pull/817, after which I can remove the note in `tests/test_pcodec.py` here.

agriyakhetarpal committed:
cibuildwheel v4.1.0 is released, build for both ABIs

agriyakhetarpal committed:
Use `before-all` stage to apply patches

agriyakhetarpal committed:
`compression.zstd` now available. Re-enable `pyzstd` tests

agriyakhetarpal committed:
Add pyzstd ignore marker for cp313

agriyakhetarpal committed:
Work around cibuildwheel shell quoting bug

agriyakhetarpal committed:
Revert "Work around cibuildwheel shell quoting bug"

This reverts commit 1e917479d9865fafc06229d25d155fae6fd89a7d.

agriyakhetarpal committed:
Test pypa/cibuildwheel#2913

agriyakhetarpal committed:
pypa/cibuildwheel#2913 is working, revert changes

This reverts commit 97e91538b2a2c177afc5dd11f938e3da259076ab.

agriyakhetarpal committed:
Reapply "Work around cibuildwheel shell quoting bug"

This reverts commit 5db34afb8d888be43f338b84746fd1a3ea693de5.

agriyakhetarpal committed:
Merge main and fix conflicts

agriyakhetarpal committed:
Merge main, pcodec is now available

agriyakhetarpal committed:
Remove outdated note about pcodec availability
